### ISSUE TITLE
Unification does not respect rewrite rules

### ISSUE THREAD

**Issue by divipp** [agda/agda#3001]
I am working on type-theory-in-type-theory, with lots of postulates and rewrite rules.
I observed a bad interaction between rewrite rules.
I could produce a minimal example with nats and fins:

```agda
{-# OPTIONS --rewriting #-}

data  _≡_ {A : Set}(x : A) : A → Set  where refl : x ≡ x

{-# BUILTIN REWRITE _≡_ #-}

postulate
  ℕ : Set
  zero : ℕ
  suc  : ℕ → ℕ

  Fin : ℕ → Set
  zeroF : (n : ℕ) → Fin (suc n)
  sucF  : (n : ℕ) → Fin n → Fin (suc n)

  collapseℕ : suc zero ≡ zero
  collapseF : (n : ℕ) → sucF (suc n) (zeroF n) ≡ zeroF (suc n)

{-# REWRITE collapseF #-}
-- {-# REWRITE collapseℕ #-} -- if I comment this out, collapseF' fails

collapseF' : sucF (suc zero) (zeroF zero) ≡ zeroF (suc zero)
collapseF' = refl
```

The problem is that if the `collapseℕ` rewrite rule is active, then the `collapseF` rewrite rule fails to fire in the type of `collapseF'`, so `refl` is rejected in its definition.

As I understood the documentation of rewrite rules, this should not happen, because the substitution which maps `n` to `zero` should be perfect for `collapseF` to fire.

**Comment by jespercockx** [comment]
Currently `--rewriting` does not give you any guarantees about the confluence of your rewrite rules. If your rewrite system is non-confluent, you may observe all kinds of weird behaviour like you do here.

In your example, the set of rewrite rules is indeed non-confluent: `sucF {suc zero} (zeroF zero)` rewrites to both to `zeroF (suc zero)` (via `collapseF`) and to `sucF {zero} (zeroF zero)` (via `collapseℕ`). The latter does not match the lhs of the `collapseF` rule because of the implicit argument {zero} to `sucF` which does not match the implicit argument `{suc n}` in the lhs.

You can add an extra rule to make it again confluent:
```
postulate completion : sucF zero (zeroF zero) ≡ zeroF (suc zero)
{-# REWRITE completion #-}
```
Alternatively, you could generalize the rewrite rule `collapseℕ` to `(n : ℕ) → suc n ≡ zero` and swapping the order in which you add the two rewrite rules.

I've been thinking about implementing a confluence checker for rewrite rules or perhaps even an automatic completion algorithm, but so far less experimental parts of Agda have consumed my attention.

**Comment by divipp** [comment]
> because of the implicit argument {zero} to `sucF` which does not match the implicit argument `{suc n}` in the lhs

I assumed that the matching algorithm works like this:

-  matching problem: `sucF {zero} (zeroF zero)`  =?  `sucF {suc n} (zeroF n)`
-  matching subproblem: `zero` =? `suc n`
    - cannot decide, postpone this subproblem 
-  matching subproblem: `zeroF zero` =? `zeroF n`
    -  matching subproblem: `zero` =? `n`
        - let `n = zero`
-  revisit `zero` =? `suc n`, where `n = zero`
    - `zero` =? `suc zero`
         - these are the same after evaluation

I admit that there is a gap between the steps  `zeroF zero` =? `zeroF n`  ===>  `zero` =? `n` because it is not known that `zeroF` is injective.
Also, maybe deciding when to postpone (e.g. postponing `zero` =? `suc n`) is not feasible.

Anyway, if matching would be smarter, then the above rewrite rules could be used together without a completion.
I am wondering that the changing the implementation of 'match' could be easier than an automating completion algorithm and a confluence checker?

**Comment by jespercockx** [comment]
You're right, I hadn't considered that this was actually due to a limitation of the current matching algorithm. I think the current algorithm may fail hard on the problem `zero =? suc n` rather than postponing it, I'll see how hard that is to change.

**Comment by jespercockx** [comment]
To increase the power of the matching algorithm in this way, we would have to reduce the lhs of the rewrite rule at the point where we apply the rule (currently we only normalize the lhs at the point where the rule is first added). One potential problem is that by reducing the lhs, we may lose bindings for the variables that are used in the rhs. 

To properly fix the algorithm to deal with these kind of problems, we would have to replace the current higher-order *matching* algorithm with full-blown higher-order *unification*, which would allow us to reorder or postpone equations and dynamically fill in the solutions for already instantiated variables. Again, not something I'm necessarily opposed to, but it seems like it could take some time to implement.

**Comment by divipp** [comment]
I updated the main example to avoid hidden arguments (sorry for the inconvenience it causes).

**Comment by AndrasKovacs** [comment]
I think the issue is that `zero =? suc n` is never postponed, because it is unsolvable without rewrite rules, but it may be solvable with rewrite rules.

Could this be solved by allowing such equations to be postponed when one of the sides has a head symbol with attached rewrite rules?

**Comment by divipp** [comment]
I think that reducing the lhs of the rewrite rule where we apply a rule is not an issue here.

Also, the documentation of `Agda.TypeChecking.Rewriting` tells that it uses a *unification* algorithm.

**Comment by divipp** [comment]
Example which shows that the problem is present with just one rewrite rule and general unification:

```agda
{-# OPTIONS --rewriting #-}

data  _≡_ {A : Set}(x : A) : A → Set  where refl : x ≡ x

{-# BUILTIN REWRITE _≡_ #-}

postulate
  ℕ : Set
  zero : ℕ
  suc  : ℕ → ℕ

data Fin : ℕ → Set where
  zeroF : (n : ℕ) → Fin (suc n)
  sucF  : (n : ℕ) → Fin n → Fin (suc n)

postulate
  collapseℕ : suc zero ≡ zero

-- {-# REWRITE collapseℕ #-}  -- unification fails if this is turned on

lhs : (n : ℕ) → Fin (suc (suc n))
lhs n = sucF (suc n) (zeroF n)

unification : Set
unification = lhs _ ≡ sucF (suc zero) (zeroF zero)
```

**Comment by divipp** [comment]
A smaller example:

```agda
{-# OPTIONS --rewriting #-}

data  _≡_ {A : Set}(x : A) : A → Set  where refl : x ≡ x

{-# BUILTIN REWRITE _≡_ #-}

postulate
  A   : Set
  a   : A
  id  : A → A
  id≡ : id a ≡ a

data T : A → Set where
  t : (a : A) → T (id a)

{-# REWRITE id≡ #-}  -- unification fails if this is active

unification : Set
unification = t a ≡ t _
```

It seems that Agda tries to fill the hole with
`t _1` : `T a`   ==>  `id _1` = `a`  ==> postpone
and does not use the identity 
`t _1` = `t a` ==> `_1` = `a`.

**Comment by jespercockx** [comment]
It's not surprising that in your last example Agda fails to find a solution, since nothing triggers unification of `t a` and `t _`. However, the following should work but doesn't
```agda
{-# OPTIONS --rewriting #-}

data  _≡_ {A : Set}(x : A) : A → Set  where refl : x ≡ x

{-# BUILTIN REWRITE _≡_ #-}

postulate
  A   : Set
  a   : A
  id  : A → A
  id≡ : id a ≡ a

data T : A → Set where
  t : (a : A) → T (id a)

{-# REWRITE id≡ #-}  -- unification fails if this is active

unification : t a ≡ t _
unification = refl
```
This seems to be a real bug in the constraint solver and not just a limitation of the current matching algorithm like in the original example.

**Comment by divipp** [comment]
Thanks for the correction of the example.

There is an increasing interest in rewriting rules here in Budapest. It would be nice if this issue was marked as a bug if you think so (I cannot change the labels).
